### PR TITLE
Office365 - Remove restriction of 1000 contacts in search

### PR DESCRIPTION
**Pull request by nissou91**

### CODE REVIEW

**Review by bloom1 — CHANGES_REQUESTED** [review]
I tested the PR for real on `wazo-test`. I added some debug lines to see what was happening and here is the output: 
```
2022-11-28 16:21:56,417 [5995] (ERROR) (wazo_dird.plugins.office365_backend.services): An error occured while fetching information from microsoft endpoint
2022-11-28 16:21:56,417 [5995] (DEBUG) (wazo_dird.plugins.office365_backend.services): Status code: 400, response: "{"error":{"code":"BadRequest","message":"Query option '$top' was specified more than once, but it must be specified at most once.","innerError":{"date":"2022-11-28T21:21:56","request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx","client-request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx"}}}"
```

File: integration_tests/suite/helpers/fixtures/http.py [hunk 1]
```diff
@@ -197,6 +197,51 @@ def wrapper(self, *args, **kwargs):
...
+                '/v1.0/me/contacts', {'@odata.count': len(pages)}, 200
+            )
+            count_expectation['httpRequest']['queryStringParameters'] = {
+                '$count': ['true'],
+            }
+
+            count_expectation['times']['unlimited'] = True
+            mock_server.mock_any_response(count_expectation)
+            counter = 0
+            while counter < len(pages):
```
[the remark this answers is not in the document]
Comment (nissou91) [reply integration_tests/suite/helpers/fixtures/http.py:218]: yes you are right; I was using it for another purpose (testing) and forgot to remove the counter; fixed it :)

File: integration_tests/suite/test_office365_contacts.py [hunk 2]
```diff
@@ -79,6 +143,31 @@ def test_unknown_source(self):
...
+        result = self.list_(self.client, self.source_uuid)
+        assert_that(
+            result,
+            has_entries(
+                total=5000,
+                filtered=5000,
+            ),
+        )
+        result_items = result.get('items')
+        for i in range(5000):
```
[the remark this answers is not in the document]
Comment (nissou91) [reply integration_tests/suite/test_office365_contacts.py:157]: it is the total number of contacts that I generate (all three pages); I will create a constant for that.

File: integration_tests/suite/test_office365_contacts.py [hunk 3]
```diff
@@ -1,7 +1,14 @@
 # Copyright 2020-2021 The Wazo Authors  (see the AUTHORS file)
```
[the remark this answers is not in the document]
Comment (nissou91) [reply integration_tests/suite/test_office365_contacts.py:1]: Thanks, I did not know about this tool; I changed them manually for now :)

File: integration_tests/suite/helpers/fixtures/http.py [hunk 4]
```diff
@@ -197,6 +197,51 @@ def wrapper(self, *args, **kwargs):
     return decorator
 
 
+def office365_paginated_result(pages):
```
[the remark this answers is not in the document]
Comment (nissou91) [reply integration_tests/suite/helpers/fixtures/http.py:200]: Done

File: wazo_dird/plugins/office365_backend/services.py [hunk 5]
```diff
@@ -35,18 +35,69 @@ def get_contacts(self, microsoft_token, url, **list_params):
...
-                logger.debug('Successfully fetched contacts from microsoft.')
-                logger.debug('Raw data: %s', response.text)
-                return response.json().get('value', [])
-            else:
-                logger.error(
-                    'An error occured while fetching information from microsoft endpoint'
-                )
-                raise UnexpectedEndpointException(
-                    endpoint=url, error_code=response.status_code
+            if count == 1:
```
[the remark this answers is not in the document]
Comment (nissou91) [reply wazo_dird/plugins/office365_backend/services.py:38]: Done

File: wazo_dird/plugins/office365_backend/services.py [hunk 6]
```diff
@@ -35,18 +35,69 @@ def get_contacts(self, microsoft_token, url, **list_params):
...
+                        next_page_response = requests.get(
+                            next_page_url,
+                            headers=headers,
+                            params={
+                                '$top': count,
+                            },
+                        )
+                        if next_page_response.status_code == 200:
+                            logger.debug(
+                                f'Successfully fetched contacts from microsoft page: {next_page_url}'
```
[the remark this answers is not in the document]
Comment (nissou91) [reply wazo_dird/plugins/office365_backend/services.py:74]: corrected

**Comment by nissou91** [comment]
> I tested the PR for real on `wazo-test`. I added some debug lines to see what was happening and here is the output:
> 
> ```
> 2022-11-28 16:21:56,417 [5995] (ERROR) (wazo_dird.plugins.office365_backend.services): An error occured while fetching information from microsoft endpoint
> 2022-11-28 16:21:56,417 [5995] (DEBUG) (wazo_dird.plugins.office365_backend.services): Status code: 400, response: "{"error":{"code":"BadRequest","message":"Query option '$top' was specified more than once, but it must be specified at most once.","innerError":{"date":"2022-11-28T21:21:56","request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx","client-request-id":"8de4869f-30c9-45b7-8441-xxxefb82exxx"}}}"
> ```

I have removed the `$top` query param from the request and the expectation to avoid redundancy error like the one you got.

**Comment by nissou91** [comment]
recheck

File: wazo_dird/plugins/office365_backend/services.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (bloom1) [reply wazo_dird/plugins/office365_backend/services.py]: Sorry if it was not clear: the formatting should not be applied yet, so the parameter should be `logger.debug('... %s', variable)`

[the remark this answers is not in the document]
Comment (nissou91) [reply wazo_dird/plugins/office365_backend/services.py]: ahh okay, got it

[the remark this answers is not in the document]
Comment (nissou91) [reply wazo_dird/plugins/office365_backend/services.py]: Done

**Review by bloom1 — CHANGES_REQUESTED** [review]
I think you may want to add more test cases that cover more scenarios.

I am getting the following error:
```
2022-12-02 15:09:39,404 [32472] (ERROR) (root): An error occured in search                                                                                                                                         
Traceback (most recent call last):                                                                                                                                                                                 
  File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/helpers.py", line 46, in execute                                                                                                                                   
    return function(*args, **kwargs)                                                                                                                                                                               
  File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/plugins/office365_backend/plugin.py", line 91, in search                                                                                                           
    sorted_contacts = sorted(filtered_contacts, key=itemgetter('givenName'))                                                                                                                                       
TypeError: '<' not supported between instances of 'str' and 'NoneType'                                                                                                                                             
2022-12-02 15:09:39,419 [32472] (INFO) (flask.app): response to 10.0.4.2 in 0.83s: GET https://wazo-test/api/dird/0.1/directories/lookup/default?term=a 200   
```
Another question: did you test the PR with a real stack and a real office365 account?

**Comment by nissou91** [comment]
> I think you may want to add more test cases that cover more scenarios.
> 
> I am getting the following error:
> 
> ```
> 2022-12-02 15:09:39,404 [32472] (ERROR) (root): An error occured in search                                                                                                                                         
> Traceback (most recent call last):                                                                                                                                                                                 
>   File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/helpers.py", line 46, in execute                                                                                                                                   
>     return function(*args, **kwargs)                                                                                                                                                                               
>   File "/usr/src/wazo/wazo-dird-WP636/wazo_dird/plugins/office365_backend/plugin.py", line 91, in search                                                                                                           
>     sorted_contacts = sorted(filtered_contacts, key=itemgetter('givenName'))                                                                                                                                       
> TypeError: '<' not supported between instances of 'str' and 'NoneType'                                                                                                                                             
> 2022-12-02 15:09:39,419 [32472] (INFO) (flask.app): response to 10.0.4.2 in 0.83s: GET https://wazo-test/api/dird/0.1/directories/lookup/default?term=a 200   
> ```
> 
> Another question: did you test the PR with a real stack and a real office365 account?

I will take a look. and yes, I tried the code on my real Office365 account; where I imported the randomly generated contacts into my account and then run the code. However, I did not run it on a wazo stack...

**Comment by nissou91** [comment]
recheck

**Review by bloom1 — APPROVED** [review]

**Review by bloom1 — APPROVED** [review]

**Pull request merged**